Code/Source/solver/mat_fun.h

```diff
@@ -43,12 +47,62 @@
 /// \todo [TODO:DaveP] this should just be a namespace?
 //
 namespace mat_fun {
+    // Helper function to convert Array<double> to Eigen::Matrix
+    template <typename MatrixType>
+    MatrixType toEigenMatrix(const Array<double>& src) {
+        MatrixType mat;
+        for (int i = 0; i < mat.rows(); ++i)
+            for (int j = 0; j < mat.cols(); ++j)
+                mat(i, j) = src(i, j);
+        return mat;
+    }
+
+    // Helper function to convert Eigen::Matrix to Array<double>
+    template <typename MatrixType>
+    void toArray(const MatrixType& mat, Array<double>& dest) {
+        for (int i = 0; i < mat.rows(); ++i)
+            for (int j = 0; j < mat.cols(); ++j)
+                dest(i, j) = mat(i, j);
+    }
+
+    // Helper function to convert a higher-dimensional array like Dm
+    template <typename MatrixType>
+    void copyDm(const MatrixType& mat, Array<double>& dest, int rows, int cols) {
+        for (int i = 0; i < rows; ++i)
+            for (int j = 0; j < cols; ++j)
+                dest(i, j) = mat(i, j);
+    }
+
+    template <int nsd>
+    Eigen::Matrix<double, nsd, 1> cross_eigen(const Eigen::Matrix<double, nsd, 1>& u, const Eigen::Matrix<double, nsd, 1>& v) {
+        if constexpr (nsd == 2) {
+            return Eigen::Matrix<double, 2, 1>(v(1), - v(0));
+        }
+        else if constexpr (nsd == 3) {
+            return u.cross(v);
+        }
+        else {
+            throw std::runtime_error("Invalid number of spatial dimensions");
+        }
+    }
+
     double mat_ddot(const Array<double>& A, const Array<double>& B, const int nd);
+    
+    template <int nsd>
+    double mat_ddot_eigen(const Eigen::Matrix<double, nsd, nsd>& A, const Eigen::Matrix<double, nsd, nsd>& B) {
```

Comment:
Rename this function to something meaningful without `mat` and `eigen`, like `double_dot_product`.